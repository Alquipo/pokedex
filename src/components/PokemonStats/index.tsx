import { useState, useEffect } from 'react'

import ProgressBarPokemon, { Sizes } from 'components/ProgressBarPokemon'

import * as S from './styles'
import { PokemonTypesProps } from 'components/PokemonCard'

export type StatsProps = {
  baseStat: number
  name: string
}

export type PokemonStatsProps = {
  stats: StatsProps[]
  types: PokemonTypesProps[]
}

const PokemonStats = ({ stats, types }: PokemonStatsProps) => {
  const [statsBase, setStatsBase] = useState(true)
  const [statsMin, setStatsMin] = useState(false)
  const [statsMax, setStatsMax] = useState(false)

  const [clickStats, setClickStats] = useState<Sizes>('base')

  useEffect(() => {
    switch (clickStats) {
      case 'base': {
        setStatsBase(true)
        setStatsMin(false)
        setStatsMax(false)

        break
      }
      case 'min': {
        setStatsMin(true)
        setStatsBase(false)
        setStatsMax(false)

        break
      }
      case 'max': {
        setStatsMax(true)
        setStatsMin(false)
        setStatsBase(false)
        break
      }
    }
  }, [clickStats])

  const totalStats = stats.reduce(
    (acumulador, stat) => acumulador + stat.baseStat,
    0
  )

  /* istanbul ignore next */
  stats.map((stat) => {
    stat.name === 'special-defense' && (stat.name = 'Sp.Defense')

    stat.name === 'special-attack' && (stat.name = 'Sp.Attack')
  })
  return (
    <S.Wrapper>
      <S.WrapperRotate>
        <S.Table>
          <tbody>
            <tr>
              <th></th>
              <th>
                <S.Button
                  typeColor={types[0]}
                  buttonActive={statsBase}
                  onClick={() => setClickStats('base')}
                  aria-label="button base"
                >
                  Base
                </S.Button>
              </th>
              <th>
                <S.Button
                  typeColor={types[0]}
                  buttonActive={statsMin}
                  onClick={() => setClickStats('min')}
                  aria-label="button min"
                >
                  Min
                </S.Button>
              </th>
              <th>
                <S.Button
                  typeColor={types[0]}
                  buttonActive={statsMax}
                  onClick={() => setClickStats('max')}
                  aria-label="button max"
                >
                  Max
                </S.Button>
              </th>
            </tr>

            {stats.map((stat, index) => (
              <tr key={index}>
                <td colSpan={1}>{stat.name}</td>

                <td colSpan={3}>
                  <ProgressBarPokemon
                    bgColor={types[0]}
                    stats={stat.baseStat}
                    size={clickStats}
                  />
                </td>
              </tr>
            ))}
            <tr>
              <td>Total</td>
              <td>{totalStats}</td>
            </tr>
          </tbody>
        </S.Table>
      </S.WrapperRotate>
    </S.Wrapper>
  )
}

export default PokemonStats
