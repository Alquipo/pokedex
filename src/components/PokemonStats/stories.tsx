import { Story, Meta } from '@storybook/react/types-6-0'
import PokemonStats, { PokemonStatsProps } from '.'

export default {
  title: 'PokemonPage/PokemonStats',
  component: PokemonStats,
  args: {
    types: ['fire', 'flying'],
    stats: [
      {
        baseStat: 100,
        name: 'hp'
      },
      {
        baseStat: 110,
        name: 'attack'
      },
      {
        baseStat: 80,
        name: 'defense'
      },
      {
        baseStat: 154,
        name: 'Sp.attack'
      },
      {
        baseStat: 90,
        name: 'Sp.defense'
      },
      {
        baseStat: 130,
        name: 'speed'
      }
    ]
  },
  argTypes: {
    types: {
      control: {
        type: 'inline-check',
        options: [
          'normal',
          'fire',
          'water',
          'electric',
          'grass',
          'ice',
          'ground',
          'flying',
          'ghost',
          'rock',
          'fighting',
          'poison',
          'psychic',
          'bug',
          'dark',
          'steel',
          'dragon',
          'fairy'
        ]
      }
    }
  }
} as Meta

export const Default: Story<PokemonStatsProps> = (args) => (
  <div
    style={{
      margin: '2.5rem auto 0 auto'
    }}
  >
    <PokemonStats {...args} />
  </div>
)
