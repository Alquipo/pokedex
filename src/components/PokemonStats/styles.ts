import styled, { css } from 'styled-components'
import { generateMedia } from 'styled-media-query'

import { PokemonTypesProps } from 'components/PokemonCard'

const customMedia = generateMedia({
  desktop: '980px'
})

type PokemonTypeColors = {
  typeColor: PokemonTypesProps
}

export const Wrapper = styled.section`
  ${({ theme }) => css`
    display: flex;

    width: 100%;
    padding-left: 2rem;
    font-size: ${theme.font.sizes.xlarge};

    ${customMedia.greaterThan('desktop')`
 perspective: 35rem;
`}

    ${customMedia.lessThan('desktop')`
 margin-top: 5rem;
`}

cursor: default;
  `}
`

export const WrapperRotate = styled.div`
  display: flex;
  justify-content: center;
  align-items: center;

  width: 100%;

  padding-right: 5rem;

  ${customMedia.greaterThan('desktop')`
      transform: rotateY(-30deg);
      transition: 0.4s;

      &:hover {
        transform: rotateY(0);
      }
    `}
`

export const Table = styled.table`
  ${({ theme }) => css`
    height: 100%;

    margin-bottom: 1rem;
    line-height: 1.5;
    color: ${theme.colors.gray};

    border-collapse: collapse;

    text-transform: capitalize;

    td {
      padding: ${theme.spacings.xsmall};
      vertical-align: center;
    }

    td:first-child {
      font-weight: ${theme.font.medium};
      text-align: right;
      width: 8rem;
      vertical-align: top;

      padding-right: 2rem;
    }
  `}
`

type ButtonProps = {
  buttonActive: boolean
} & PokemonTypeColors

export const Button = styled.button<ButtonProps>`
  ${({ theme, typeColor, buttonActive }) => css`
    border-radius: 5rem;
    font-weight: ${theme.font.bold};
    text-align: center;
    font-size: 1.6rem;
    line-height: 1.7;

    border: 1px solid transparent;
    padding: 0.4rem 0.8rem;

    color: ${theme.colors.gray};
    background-color: ${theme.colors.lightBg};

    ${buttonActive &&
    css`
      color: ${theme.colors.white};
      background-color: ${theme.colors.pokemonText[typeColor]};
      transition: background-color 1s ease-in-out, width 1.5s ease-in-out,
        color 0.5s ease-in;
    `}

    &:focus {
      outline: 0;
    }
  `}
`
