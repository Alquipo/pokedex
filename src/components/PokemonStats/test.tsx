import { fireEvent, screen } from '@testing-library/react'
import { renderWithTheme } from 'utils/tests/helpers'

import PokemonStats, { PokemonStatsProps } from '.'

import mockPokemon from 'templates/Pokemon/mock'

const props: PokemonStatsProps = {
  stats: mockPokemon.stats,
  types: ['fire', 'flying']
}

jest.mock('components/ProgressBarPokemon', () => ({
  __esModule: true,
  default: function Mock({ children }: { children: React.ReactNode }) {
    return <img data-testid="Mock ProgressBarPokemon">{children}</img>
  }
}))

describe('<PokemonStats />', () => {
  it('should render the PokemonStats correctly', () => {
    const { container } = renderWithTheme(<PokemonStats {...props} />)

    expect(screen.getByRole('cell', { name: /hp/i })).toBeInTheDocument()
    expect(screen.getByRole('cell', { name: 'attack' })).toBeInTheDocument()
    expect(screen.getByRole('cell', { name: 'defense' })).toBeInTheDocument()
    expect(screen.getByRole('cell', { name: 'Sp.attack' })).toBeInTheDocument()
    expect(screen.getByRole('cell', { name: 'Sp.defense' })).toBeInTheDocument()
    expect(screen.getByRole('cell', { name: /speed/i })).toBeInTheDocument()
    expect(screen.getByRole('cell', { name: /Total/i })).toBeInTheDocument()

    expect(screen.getAllByTestId('Mock ProgressBarPokemon')).toHaveLength(6)

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should handle switch stats or button clicked', () => {
    renderWithTheme(<PokemonStats {...props} />)

    fireEvent.click(screen.getByRole('button', { name: /button min/i }))
    fireEvent.click(screen.getByRole('button', { name: /button base/i }))
    fireEvent.click(screen.getByRole('button', { name: /button max/i }))
  })
})
