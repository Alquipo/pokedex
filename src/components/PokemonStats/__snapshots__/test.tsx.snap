// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonStats /> should render the PokemonStats correctly 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  padding-left: 2rem;
  font-size: 2.0rem;
  cursor: default;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 100%;
  padding-right: 5rem;
}

.c2 {
  height: 100%;
  margin-bottom: 1rem;
  line-height: 1.5;
  color: #505050;
  border-collapse: collapse;
  text-transform: capitalize;
}

.c2 td {
  padding: 0.8rem;
  vertical-align: center;
}

.c2 td:first-child {
  font-weight: 600;
  text-align: right;
  width: 8rem;
  vertical-align: top;
  padding-right: 2rem;
}

.c3 {
  border-radius: 5rem;
  font-weight: 700;
  text-align: center;
  font-size: 1.6rem;
  line-height: 1.7;
  border: 1px solid transparent;
  padding: 0.4rem 0.8rem;
  color: #505050;
  background-color: #f2f3f5;
  color: #fff;
  background-color: #ff8a80;
  -webkit-transition: background-color 1s ease-in-out,width 1.5s ease-in-out,color 0.5s ease-in;
  transition: background-color 1s ease-in-out,width 1.5s ease-in-out,color 0.5s ease-in;
}

.c3:focus {
  outline: 0;
}

.c4 {
  border-radius: 5rem;
  font-weight: 700;
  text-align: center;
  font-size: 1.6rem;
  line-height: 1.7;
  border: 1px solid transparent;
  padding: 0.4rem 0.8rem;
  color: #505050;
  background-color: #f2f3f5;
}

.c4:focus {
  outline: 0;
}

@media (min-width:980px) {
  .c0 {
    -webkit-perspective: 35rem;
    -moz-perspective: 35rem;
    -ms-perspective: 35rem;
    perspective: 35rem;
  }
}

@media (max-width:980px) {
  .c0 {
    margin-top: 5rem;
  }
}

@media (min-width:980px) {
  .c1 {
    -webkit-transform: rotateY(-30deg);
    -ms-transform: rotateY(-30deg);
    transform: rotateY(-30deg);
    -webkit-transition: 0.4s;
    transition: 0.4s;
  }

  .c1:hover {
    -webkit-transform: rotateY(0);
    -ms-transform: rotateY(0);
    transform: rotateY(0);
  }
}

<section
  class="c0"
>
  <div
    class="c1"
  >
    <table
      class="c2"
    >
      <tbody>
        <tr>
          <th />
          <th>
            <button
              aria-label="button base"
              class="c3"
            >
              Base
            </button>
          </th>
          <th>
            <button
              aria-label="button min"
              class="c4"
            >
              Min
            </button>
          </th>
          <th>
            <button
              aria-label="button max"
              class="c4"
            >
              Max
            </button>
          </th>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            hp
          </td>
          <td
            colspan="3"
          >
            <img
              data-testid="Mock ProgressBarPokemon"
            />
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            attack
          </td>
          <td
            colspan="3"
          >
            <img
              data-testid="Mock ProgressBarPokemon"
            />
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            defense
          </td>
          <td
            colspan="3"
          >
            <img
              data-testid="Mock ProgressBarPokemon"
            />
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Sp.attack
          </td>
          <td
            colspan="3"
          >
            <img
              data-testid="Mock ProgressBarPokemon"
            />
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            Sp.defense
          </td>
          <td
            colspan="3"
          >
            <img
              data-testid="Mock ProgressBarPokemon"
            />
          </td>
        </tr>
        <tr>
          <td
            colspan="1"
          >
            speed
          </td>
          <td
            colspan="3"
          >
            <img
              data-testid="Mock ProgressBarPokemon"
            />
          </td>
        </tr>
        <tr>
          <td>
            Total
          </td>
          <td>
            664
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</section>
`;
